sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala

```diff
@@ -17,22 +17,33 @@
 
 package org.apache.spark.sql.execution.datasources.v2
 
-import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.catalyst.analysis.{AnsiTypeCoercion, TypeCoercion}
+import org.apache.spark.sql.catalyst.expressions.{Expression, Literal, SortOrder, TransformExpression, V2ExpressionUtils}
 import org.apache.spark.sql.catalyst.expressions.V2ExpressionUtils._
 import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, RebalancePartitions, RepartitionByExpression, Sort}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.connector.catalog.FunctionCatalog
+import org.apache.spark.sql.connector.catalog.functions.ScalarFunction
 import org.apache.spark.sql.connector.distributions._
 import org.apache.spark.sql.connector.write.{RequiresDistributionAndOrdering, Write}
 import org.apache.spark.sql.errors.QueryCompilationErrors
 
 object DistributionAndOrderingUtils {
 
-  def prepareQuery(write: Write, query: LogicalPlan): LogicalPlan = write match {
+  def prepareQuery(
```

Comment:
Hmm I wonder how does the write work with transforms such as bucket. For example, suppose the required distribution is `bucket(col, 100)`, Spark currently will compute the partition (bucket) ID by `murmur_hash(bucket(col, 100)) pmod 100`, so the value of `col` is essentially hashed twice. I'm not sure whether this breaks any assumption from the V2 data source side, or whether it has any effect in the hash key distributions.

Reply:
> Spark currently will compute the partition (bucket) ID by `murmur_hash(bucket(col, 100)) pmod 100`

It's only happening on the V1 write path. In V2, the bucket is resolved as `BucketTransform` in analyzing phase, and converted to evaluable catalyst expression `ApplyFunctionExpression`/`Invoke`/`StaticInvoke` here, so I don't think your concern will happen.

Reply:
Hmm, yes it will be converted to `ApplyFunctionExpression` etc. But in V2 write path, these partition transforms will be used in `RepartitionByExpression`, and then converted into `ShuffleExchangeExec` during physical planning, right? 

Checking the code path, the `ApplyFunctionExpression` / `Invoke` / `StaticInvoke`, etc, will be passed to `RepartitionByExpression` first, as field `partitionExpressions`, and then be passed to `HashPartitioning`, `RangePartitioning` etc, in `HasPartitionExpressions`, and eventually be used in places like `HashPartitioning.partitionIdExpression` where the computation of partition ID I mentioned above happened.

Reply:
Thanks for explaining, educated. After reading the code, I think you are right that "the value of `col` is essentially hashed twice", but I don't think it will bring correctness issues, because it still guarantees that the same bucket value clustered into the same partition. 
One example is Hive bucket. In `V1WritesUtils#getWriterBucketSpec`, both `HiveHash`  and `HashPartitioning#partitionIdExpression` can be used to construct `bucketIdExpression`.

Reply:
I'm just thinking whether we should enforce stronger semantics on the bucketing. Consider we build bucketed table support for DSv2 file source on top on this mechanism, does it mean for a particular row, it could be hashed to a different bucket ID than V1? What if someone wants to first write a bucketed table using V2 and then read back in V1, and perhaps do bucket join with another V1 bucketed table? I wonder if that could cause incorrect results.

Reply:
In V2, the V1 bucketed table is a partitioned table w/ a special `bucket` function (we should give it a special canonical name), `TransformExpression#isSameFunction` will forbid bucket join if they use different bucket functions. So I think there is no correctness issue.

Reply:
But in that case the V1 and V2 bucket tables are incompatible, which is not very desirable right?

Reply:
Given that different storage system usually defines their own hash functions and sharding(bucketing) rules, I don't think we can make bucket tables of all data sources fully compatible w/ each other, e.g. Spark, Hive, Iceberg using the total different hash algorithm for bucketing.

We can make that DS V1 file bucket table compatible w/ V2 file bucket table, and Hive bucket table compatible w/ V2 Hive bucket table, but DS file bucket table can not be compatible w/ Hive(neither V1 nor V2) since they use the different hash algorithm.

In detail, as `V2SessionCatalog` extends `FunctionCatalog`, to make V2 file bucket tables compatible w/ V1, we can introduce and register a `SparkBucket` V2 function using the same hash algorithm V1 in `V2SessionCatalog`. For Hive tables, just register a similar `HiveBucket` V2 function using the Hive hash algorithm in `HiveCatalog`.

Reply:
I think @sunchao brings a valid point that is easy to overlook. We have to make sure Spark writes to Hive tables in the same way no matter whether the v1 or v2 path is being used.

Would it be correct to say we have this issue because `partitionIdExpression` in `HashPartitioning` is used both for generating bucket IDs in Hive tables as well as for producing partition IDs for writing tasks? Can we use different mechanisms?

Reply:
After SPARK-32709(https://github.com/apache/spark/pull/33432), it allows using different hash expressions for generating partition IDs and bucket IDs.

Reply:
> Consider we build bucketed table support for DSv2 file source on top on this mechanism, does it mean for a particular row, it could be hashed to a different bucket ID than V1? What if someone wants to first write a bucketed table using V2 and then read back in V1, and perhaps do bucket join with another V1 bucketed table?

@sunchao, coming back to the use case you mentioned above. I think the bucket ID will be always the same as long as the task writer respects the table bucket spec and Spark shuffles all records that are supposed to land in one bucket to one task. Like @pan3793 said, I suppose V2 file source will request a distribution using a V2 function that would wrap the internal Spark hash function. That should guarantee that all records for the same bucket will land in one task. As long as the task writer uses the correct bucket expression (based on the table definition), we should be good, right?

You are right the partition/task ID during writes for a particular row may be different in V1 and V2 because of double hashing. But does it actually matter, though?

Reply:
Sorry for the late reply here. Yes, I think it should be OK as long as the task writer uses the bucket expression to decide which bucket the input record should go. There will be some extra work to achieve V1 and V2 compatibility but it's not that relevant to this PR now.